### PR TITLE
Add Recall Tests

### PR DESCRIPTION
**Pull request by naveentatikonda**
Signed-off-by: Naveen Tatikonda <navtat@amazon.com>

### Description
Adding recall tests to integration tests and backward compatibility tests. This includes two separate tests where the vectors are generated using some standard function in one test and randomly generated vectors in other test. 'l2' is the spaceType used to measure distance between two points.
 
### Issues Resolved
https://github.com/opensearch-project/k-NN/issues/250
 
### Check List
- [x] Commits are signed as per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/test/java/org/opensearch/knn/recall/RecallTestsIT.java [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+                "POST",
+                 "/_bulk"
+        );
+
+        request.addParameter("refresh", "true");
+        String body = "";
+
+        for(int i = 0; i < docCount; i++){
+          body += "{ \"index\" : { \"_index\" : \"" + index + "\", \"_id\" : \"" + String.valueOf(i+1) + "\" } }\n"
+          +"{ \"" + fieldName + "\" : "+ Arrays.toString(indexVectors[i]) +" }\n";
```
Comment (martin-gaievski) [inline src/test/java/org/opensearch/knn/recall/RecallTestsIT.java:90]: Using a StringBuilder and append should save some resources, with current implementation there are too many string objects created during string concatenations

File: src/test/java/org/opensearch/knn/recall/RecallTestsIT.java [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+    }
+
+    public double calculateRecallValue(float[][] queryVectors, List<List<float[]>> searchResults, float[][] groundTruthValues, int k, SpaceType spaceType){
+        ArrayList<Float> recalls = new ArrayList<>();
+        float dist = 0.0F;
+
+        for(int i = 0; i < queryVectors.length; i++){
+            float recallVal = 0.0F;
+            for(int j = 0; j < searchResults.get(i).size(); j++){
+                if(spaceType.getValue().equals("l2")){
```
Comment (martin-gaievski) [inline src/test/java/org/opensearch/knn/recall/RecallTestsIT.java:139]: nit: please consider additional null check and order "constant equals variable"

File: src/test/java/org/opensearch/knn/recall/RecallTestsIT.java [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+            recalls.add(recallVal / k);
+        }
+
+        double sum = recalls.stream().reduce((a,b)->a+b).get();
+        return sum/recalls.size();
+    }
+
+    public float[][] getIndexVectorsFromIndex(String testIndex, String testField, int docCount, int dimensions) throws IOException {
+        float[][] vectors = new float[docCount][dimensions];
+        int i = 0;
```
Comment (martin-gaievski) [inline src/test/java/org/opensearch/knn/recall/RecallTestsIT.java:156]: could you please move initialization of variable "i" closer to the loop where it's used, otherwise code is a bit hard to read

File: src/test/java/org/opensearch/knn/TestUtils.java [hunk 4]
```diff
@@ -15,14 +15,12 @@
...
 import org.opensearch.knn.index.codec.KNNCodecUtil;
-
-import java.io.BufferedReader;
-import java.io.FileReader;
-import java.io.IOException;
-import java.util.ArrayList;
-import java.util.List;
-import java.util.Map;
+import org.opensearch.knn.plugin.script.KNNScoringUtil;
+import java.io.*;
```
Comment (jmazanec15) [inline src/test/java/org/opensearch/knn/TestUtils.java:21]: There is some debate around whether wildcard imports should be used. In general, I think we should avoid them. Google style follows this convention as well: https://google.github.io/styleguide/javaguide.html#s3.3.1-wildcard-imports. 

Here is some discussion: https://stackoverflow.com/questions/147454/why-is-using-a-wild-card-with-a-java-import-statement-bad

File: src/test/java/org/opensearch/knn/TestUtils.java [hunk 5]
```diff
@@ -32,6 +30,62 @@
...
+        }
+        return Vectors;
+    }
+
+    public static float[][] computeGroundTruthDistances(float[][] indexVectors, float[][] queryVectors, SpaceType spaceType, int k){
+        float[][] groundTruthValues = new float[queryVectors.length][k];
+        for(int i = 0; i < queryVectors.length; i++){
+            float[] distValues = new float[indexVectors.length];
+            for(int j = 0; j < indexVectors.length; j++){
+                if(spaceType.getValue().equals("l2"))
```
[the remark this answers is not in the document]
Comment (jmazanec15) [reply src/test/java/org/opensearch/knn/TestUtils.java:80]: +1

File: src/test/java/org/opensearch/knn/TestUtils.java [hunk 6]
```diff
@@ -32,6 +30,62 @@
...
+                m=m+5;
+                n=n+7;
+            }
+            Vectors[i] = vector;
+        }
+        return Vectors;
+    }
+
+    public static float[][] generateRandomVectors(int numVectors, int dimensions){
+        float[][] Vectors = new float[numVectors][dimensions];
```
Comment (jmazanec15) [inline src/test/java/org/opensearch/knn/TestUtils.java:63]: Why is this var capitalized?

File: src/test/java/org/opensearch/knn/TestUtils.java [hunk 7]
```diff
@@ -32,6 +30,62 @@
...
+            float[] vector = new float[dimensions];
+            for (int j = 0; j < dimensions; j++) {
+                vector[j] = (float) ((j * j * p++) - (i * i * q--));
+            }
+            Vectors[i] = vector;
+        }
+        return Vectors;
+    }
+
+    public static float[][] generateQueryVectors(int numVectors, int dimensions){
```
Comment (jmazanec15) [inline src/test/java/org/opensearch/knn/TestUtils.java:47]: What makes these vectors query vectors?

File: src/test/java/org/opensearch/knn/TestUtils.java [hunk 8]
```diff
@@ -32,6 +30,62 @@
     public static final String BWCSUITE_ROUND = "tests.rest.bwcsuite_round";
     public static final String TEST_CLUSTER_NAME = "tests.clustername";
 
+    public static float[][] generateVectors(int numVectors, int dimensions) {
```
Comment (jmazanec15) [inline src/test/java/org/opensearch/knn/TestUtils.java:33]: I think generateVectors and generateQueryVectors function names should be refactored. By the name, it is not clear what makes them different. generateRandomVectors is descriptive in that it says how it is generating the vectors.

File: src/test/java/org/opensearch/knn/TestUtils.java [hunk 9]
```diff
@@ -32,6 +30,62 @@
...
+            for (int j = 0; j < dimensions; j++) {
+                vector[j] = random().nextFloat();
+            }
+            Vectors[i] = vector;
+        }
+        return Vectors;
+    }
+
+    public static float[][] computeGroundTruthDistances(float[][] indexVectors, float[][] queryVectors, SpaceType spaceType, int k){
+        float[][] groundTruthValues = new float[queryVectors.length][k];
```
Comment (jmazanec15) [inline src/test/java/org/opensearch/knn/TestUtils.java:76]: This function could use a long comment about what it is doing.

File: src/test/java/org/opensearch/knn/bwc/KNNBackwardsCompatibilityIT.java [hunk 10]
```diff
@@ -15,17 +17,26 @@
...
 
 import static org.opensearch.knn.TestUtils.*;
 
 public class KNNBackwardsCompatibilityIT extends KNNRestTestCase {
     private static final String CLUSTER_NAME = System.getProperty(TEST_CLUSTER_NAME);
     private final String testIndexName = KNN_BWC_PREFIX+"test-index";
+    private final String testIndex_Recall = KNN_BWC_PREFIX+"test-index-recall";
     private final String testFieldName = "test-field";
+    private final String testField_Recall = "test-field-recall";
+    private final String testIndex_Recall_Old = KNN_BWC_PREFIX+"test-index-recall-value-old";
```
Comment (jmazanec15) [inline src/test/java/org/opensearch/knn/bwc/KNNBackwardsCompatibilityIT.java:33]: Add space between + and operands

File: src/test/java/org/opensearch/knn/bwc/KNNBackwardsCompatibilityIT.java [hunk 11]
```diff
@@ -82,6 +93,7 @@ private ClusterType getClusterType(){
     @SuppressWarnings("unchecked")
     public void testBackwardsCompatibility() throws Exception {
         String uri = getUri(getClusterType());
+        RecallTestsIT recallTestsIT = new RecallTestsIT();
```
Comment (VijayanB) [inline src/test/java/org/opensearch/knn/bwc/KNNBackwardsCompatibilityIT.java:96]: I am not sure whether you can use one test method in another test

File: src/test/java/org/opensearch/knn/TestUtils.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (naveentatikonda) [reply src/test/java/org/opensearch/knn/TestUtils.java]: updated

File: src/test/java/org/opensearch/knn/bwc/KNNBackwardsCompatibilityIT.java [hunk 12]
```diff
@@ -138,7 +155,16 @@ else if(round.equals("second")) {
                     }
                     else {
                         assertEquals("4", results.get(0).getDocId());
+                    }
+
+                    double recallValMixed = recallTestsIT.getRecallValue(testIndex_Recall, testField_Recall, docCount, queryCount, k_Recall, dimensions_Recall, SpaceType.L2, true);
+                    float[][] expRecallVal = recallTestsIT.getIndexVectorsFromIndex(testIndex_Recall_Old, testFieldName, 1, dimensions_Recall_Old);
+                    assertEquals(expRecallVal[0][0], recallValMixed, 0.2);
+
+                    if(round.equals("third")){
```
[the remark this answers is not in the document]
Comment (naveentatikonda) [reply src/test/java/org/opensearch/knn/bwc/KNNBackwardsCompatibilityIT.java]: updated

File: src/test/java/org/opensearch/knn/recall/RecallTestsIT.java (hunk 1 above)
Comment (naveentatikonda) [reply src/test/java/org/opensearch/knn/recall/RecallTestsIT.java]: updated

File: src/test/java/org/opensearch/knn/recall/RecallTestsIT.java (hunk 2 above)
Comment (naveentatikonda) [reply src/test/java/org/opensearch/knn/recall/RecallTestsIT.java]: updated

File: src/test/java/org/opensearch/knn/recall/RecallTestsIT.java (hunk 3 above)
Comment (naveentatikonda) [reply src/test/java/org/opensearch/knn/recall/RecallTestsIT.java]: updated

File: src/test/java/org/opensearch/knn/TestUtils.java (hunk 4 above)
Comment (naveentatikonda) [reply src/test/java/org/opensearch/knn/TestUtils.java]: I was not aware of this. Thanks for sharing it. updated the code.

File: src/test/java/org/opensearch/knn/TestUtils.java (hunk 6 above)
Comment (naveentatikonda) [reply src/test/java/org/opensearch/knn/TestUtils.java]: updated

File: src/test/java/org/opensearch/knn/TestUtils.java (hunk 7 above)
Comment (naveentatikonda) [reply src/test/java/org/opensearch/knn/TestUtils.java]: changed the function name and logic

File: src/test/java/org/opensearch/knn/TestUtils.java (hunk 8 above)
Comment (naveentatikonda) [reply src/test/java/org/opensearch/knn/TestUtils.java]: updated

File: src/test/java/org/opensearch/knn/TestUtils.java (hunk 9 above)
Comment (naveentatikonda) [reply src/test/java/org/opensearch/knn/TestUtils.java:86]: added the comment

File: src/test/java/org/opensearch/knn/bwc/KNNBackwardsCompatibilityIT.java (hunk 10 above)
Comment (naveentatikonda) [reply src/test/java/org/opensearch/knn/bwc/KNNBackwardsCompatibilityIT.java]: updated

File: src/test/java/org/opensearch/knn/TestUtils.java [hunk 13]
```diff
@@ -32,6 +30,62 @@
...
+
+    public static float[][] generateQueryVectors(int numVectors, int dimensions){
+        float[][] Vectors = new float[numVectors][dimensions];
+        int m = 1000, n = 250;
+        for (int i = 0; i < numVectors; i++) {
+            float[] vector = new float[dimensions];
+            for (int j = 0; j < dimensions; j++) {
+                vector[j] = (float)((i * m++) - (j * n--));
+                m=m+5;
+                n=n+7;
```
[the remark this answers is not in the document]
Comment (naveentatikonda) [reply src/test/java/org/opensearch/knn/TestUtils.java]: That number was actually picked randomly. But now, I have updated the logic by generating the vectors using random function with a seed. So that, it avoids all this confusion and serves the purpose of generating same vectors for each run(standard vectors).

File: src/test/java/org/opensearch/knn/bwc/KNNBackwardsCompatibilityIT.java (hunk 11 above)
Comment (naveentatikonda) [reply src/test/java/org/opensearch/knn/bwc/KNNBackwardsCompatibilityIT.java]: refactored the code

File: src/test/java/org/opensearch/knn/bwc/KNNBackwardsCompatibilityIT.java [hunk 14]
```diff
@@ -5,27 +5,44 @@
...
 import org.opensearch.client.Request;
 import org.opensearch.client.Response;
 import org.junit.Assert;
 import org.opensearch.common.settings.Settings;
+import org.opensearch.knn.index.SpaceType;
 import org.opensearch.test.rest.OpenSearchRestTestCase;
-
-import java.lang.*;
-
+import org.opensearch.knn.recall.RecallTestsIT;
```
Comment (martin-gaievski) [inline src/test/java/org/opensearch/knn/bwc/KNNBackwardsCompatibilityIT.java:28]: why we need this import?

File: src/test/java/org/opensearch/knn/KNNRestTestCase.java [hunk 15]
```diff
@@ -782,6 +783,57 @@ public void bulkIngestRandomVectors(String indexName, String fieldName, int numV
...
+                "POST",
+                "/_bulk"
+        );
+
+        request.addParameter("refresh", "true");
+        StringBuilder sb = new StringBuilder();
+
+        for(int i = 0; i < docCount; i++){
+            sb.append("{ \"index\" : { \"_index\" : \"" + index + "\", \"_id\" : \"" + String.valueOf(i+1) + "\" } }\n"
+                    +"{ \"" + fieldName + "\" : "+ Arrays.toString(indexVectors[i]) +" }\n");
```
Comment (martin-gaievski) [inline src/test/java/org/opensearch/knn/KNNRestTestCase.java:797]: to utilize StringBuilder features strings need to be added via append() method, something like 
sb.append("{ \"index\" : { \"_index\" : \"")
     .append(index)
     .append("\", \"_id\" : \"")
     .append(i + 1)
.....
otherwise each concatenation with "+" creates new String object

**Review by martin-gaievski — CHANGES_REQUESTED** [review]

File: src/test/java/org/opensearch/knn/bwc/KNNBackwardsCompatibilityIT.java (hunk 14 above)
Comment (naveentatikonda) [reply src/test/java/org/opensearch/knn/bwc/KNNBackwardsCompatibilityIT.java]: Sry, forgot to remove it

File: src/test/java/org/opensearch/knn/KNNRestTestCase.java (hunk 15 above)
Comment (naveentatikonda) [reply src/test/java/org/opensearch/knn/KNNRestTestCase.java]: Got it, updated

**Review by martin-gaievski — APPROVED** [review]
Lgtm, thx!

File: src/test/java/org/opensearch/knn/recall/RecallTestsIT.java [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+            kVectors = new ArrayList<>();
+            for(KNNResult result : resultsRecall){
+                kVectors.add(Floats.toArray(Arrays.stream(result.getVector()).collect(Collectors.toList())));
+            }
+            searchResults.add(kVectors);
+        }
+
+        return searchResults;
+    }
+}
```
Comment (VijayanB) [inline src/test/java/org/opensearch/knn/recall/RecallTestsIT.java:91]: nit: and new line

File: src/test/java/org/opensearch/knn/TestUtils.java [hunk 17]
```diff
@@ -32,6 +37,105 @@
     public static final String BWCSUITE_ROUND = "tests.rest.bwcsuite_round";
     public static final String TEST_CLUSTER_NAME = "tests.clustername";
 
+    //Generating index vectors using random function with a seed which makes these vectors standard and generate same vectors for each run.
+    public static float[][] randomlyGenerateStandardIndexVectors(int numVectors, int dimensions) {
```
Comment (VijayanB) [inline src/test/java/org/opensearch/knn/TestUtils.java:41]: nit
```suggestion
    public static float[][] generateRandomStandardIndexVectors(int numVectors, int dimensions) {
```

File: src/test/java/org/opensearch/knn/TestUtils.java [hunk 18]
```diff
@@ -32,6 +37,105 @@
...
+                Random random = new Random(seedVal++);
+                vector[j] = random.nextFloat();
+            }
+            standardIndexVectors[i] = vector;
+        }
+        return standardIndexVectors;
+    }
+
+    //Generating query vectors using random function with a seed which makes these vectors standard and generate same vectors for each run.
+    public static float[][] randomlyGenerateStandardQueryVectors(int numVectors, int dimensions){
```
Comment (VijayanB) [inline src/test/java/org/opensearch/knn/TestUtils.java:57]: nit: ```suggestion
    public static float[][] generateRandomStandardQueryVectors(int numVectors, int dimensions){
```

File: src/test/java/org/opensearch/knn/TestUtils.java [hunk 19]
```diff
@@ -32,6 +37,105 @@
...
+    }
+
+    // Here, for a given space type we will compute the 'k' shortest distances among all the index vectors for each and every query vector.
+    // These computed distances are later used while calculating Recall value to compare with the 'k' index vectors obtained for each and every search query performed.
+    public static float[][] computeGroundTruthDistances(float[][] indexVectors, float[][] queryVectors, SpaceType spaceType, int k){
+        float[][] groundTruthValues = new float[queryVectors.length][k];
+        for(int i = 0; i < queryVectors.length; i++){
+            float[] distValues = new float[indexVectors.length];
+            for(int j = 0; j < indexVectors.length; j++){
+                if(spaceType != null && ("l2").equals(spaceType.getValue()))
```
Comment (VijayanB) [inline src/test/java/org/opensearch/knn/TestUtils.java:92]:
```suggestion
                if(spaceType != null && "l2".equals(spaceType.getValue()))
```

File: src/test/java/org/opensearch/knn/TestUtils.java (hunk 17 above)
Comment (naveentatikonda) [reply src/test/java/org/opensearch/knn/TestUtils.java:41]: I felt the method name, "generateRandomStandard...." would create confusion, that's the reason I used "randomlyGenerateStandard...."

File: src/test/java/org/opensearch/knn/TestUtils.java (hunk 19 above)
Comment (naveentatikonda) [reply src/test/java/org/opensearch/knn/TestUtils.java:92]: Ack

**Review by martin-gaievski — APPROVED** [review]

File: src/test/java/org/opensearch/knn/TestUtils.java [hunk 20]
```diff
@@ -32,6 +37,105 @@
...
+        return standardQueryVectors;
+    }
+
+    public static float[][] generateRandomVectors(int numVectors, int dimensions){
+        float[][] randomVectors = new float[numVectors][dimensions];
+
+        for (int i = 0; i < numVectors; i++) {
+            float[] vector = new float[dimensions];
+            for (int j = 0; j < dimensions; j++) {
+                vector[j] = random().nextFloat();
```
[the remark this answers is not in the document]
Comment (naveentatikonda) [reply src/test/java/org/opensearch/knn/TestUtils.java:78]: In the existing code base, random() from LuceneTestCase is used. So, I tried to use it to generate random vectors but I don't think there is an option to seed it. So, I used the random function with seed to generate Standard vectors

File: src/test/java/org/opensearch/knn/TestUtils.java [hunk 21]
```diff
@@ -32,6 +37,105 @@
...
+    }
+
+    //Generating query vectors using random function with a seed which makes these vectors standard and generate same vectors for each run.
+    public static float[][] randomlyGenerateStandardQueryVectors(int numVectors, int dimensions){
+        float[][] standardQueryVectors = new float[numVectors][dimensions];
+        int seedVal = 1;
+
+        for (int i = 0; i < numVectors; i++) {
+            float[] vector = new float[dimensions];
+            for (int j = 0; j < dimensions; j++) {
```
[the remark this answers is not in the document]
Comment (naveentatikonda) [reply src/test/java/org/opensearch/knn/TestUtils.java:63]: Ack

File: src/test/java/org/opensearch/knn/TestUtils.java [hunk 22]
```diff
@@ -32,6 +37,105 @@
     public static final String BWCSUITE_ROUND = "tests.rest.bwcsuite_round";
     public static final String TEST_CLUSTER_NAME = "tests.clustername";
 
+    //Generating index vectors using random function with a seed which makes these vectors standard and generate same vectors for each run.
+    public static float[][] randomlyGenerateStandardIndexVectors(int numVectors, int dimensions) {
+        float[][] standardIndexVectors = new float[numVectors][dimensions];
+        int seedVal = 500;
```
[the remark this answers is not in the document]
Comment (naveentatikonda) [reply src/test/java/org/opensearch/knn/TestUtils.java:43]: Ack

File: src/test/java/org/opensearch/knn/TestUtils.java [hunk 23]
```diff
@@ -32,6 +37,105 @@
...
+
+        for (int i = 0; i < numVectors; i++) {
+            float[] vector = new float[dimensions];
+            for (int j = 0; j < dimensions; j++) {
+                Random random = new Random(seedVal++);
```
[the remark this answers is not in the document]
Comment (naveentatikonda) [reply src/test/java/org/opensearch/knn/TestUtils.java:48]: Ack

File: src/test/java/org/opensearch/knn/TestUtils.java [hunk 24]
```diff
@@ -32,6 +37,105 @@
...
+            randomVectors[i] = vector;
+        }
+        return randomVectors;
+    }
+
+    // Here, for a given space type we will compute the 'k' shortest distances among all the index vectors for each and every query vector.
+    // These computed distances are later used while calculating Recall value to compare with the 'k' index vectors obtained for each and every search query performed.
+    public static float[][] computeGroundTruthDistances(float[][] indexVectors, float[][] queryVectors, SpaceType spaceType, int k){
+        float[][] groundTruthValues = new float[queryVectors.length][k];
+        for(int i = 0; i < queryVectors.length; i++){
```
[the remark this answers is not in the document]
Comment (naveentatikonda) [reply src/test/java/org/opensearch/knn/TestUtils.java:89]: Ack

File: src/test/java/org/opensearch/knn/TestUtils.java [hunk 25]
```diff
@@ -32,6 +37,105 @@
...
+            }
+            randomVectors[i] = vector;
+        }
+        return randomVectors;
+    }
+
+    // Here, for a given space type we will compute the 'k' shortest distances among all the index vectors for each and every query vector.
+    // These computed distances are later used while calculating Recall value to compare with the 'k' index vectors obtained for each and every search query performed.
+    public static float[][] computeGroundTruthDistances(float[][] indexVectors, float[][] queryVectors, SpaceType spaceType, int k){
+        float[][] groundTruthValues = new float[queryVectors.length][k];
```
[the remark this answers is not in the document]
Comment (naveentatikonda) [reply src/test/java/org/opensearch/knn/TestUtils.java:88]: In this approach, we don't need to know the docIds or index Vectors of the shortest k distances computed and stored in GTValues because if the k resulting index vectors obtained from each search query gives the same l2 shortest distances in GT values, then recall value will be 1 else it will be less than 1.

It is a very good catch about using docIds and I agree with you. Will update code.

File: src/test/java/org/opensearch/knn/TestUtils.java [hunk 26]
```diff
@@ -32,6 +37,105 @@
...
+        for(int i = 0; i < queryVectors.length; i++){
+            float[] distValues = new float[indexVectors.length];
```
[the remark this answers is not in the document]
Comment (naveentatikonda) [reply src/test/java/org/opensearch/knn/TestUtils.java:90]: Ack

File: src/test/java/org/opensearch/knn/TestUtils.java [hunk 27]
```diff
@@ -32,6 +37,105 @@
...
+            for(int j = 0; j < indexVectors.length; j++){
+                if(spaceType != null && "l2".equals(spaceType.getValue()))
```
[the remark this answers is not in the document]
Comment (naveentatikonda) [reply src/test/java/org/opensearch/knn/TestUtils.java:92]: Ack

File: src/test/java/org/opensearch/knn/TestUtils.java [hunk 28]
```diff
@@ -32,6 +37,105 @@
...
+    }
+
+    public static float[][] getIndexVectors(int docCount, int dimensions, boolean isStandard){
+        if(isStandard)
+            return randomlyGenerateStandardIndexVectors(docCount, dimensions);
+        else
+            return generateRandomVectors(docCount, dimensions);
+    }
+
+    //Computing recall value by computing distances for the obtained query vectors and search results based on the given spaceType and comparing those distances with the
```
[the remark this answers is not in the document]
Comment (naveentatikonda) [reply src/test/java/org/opensearch/knn/TestUtils.java:115]: Ack

File: src/test/java/org/opensearch/knn/TestUtils.java [hunk 29]
```diff
@@ -32,6 +37,105 @@
...
+    //computed ground truth values
+    public static double calculateRecallValue(float[][] queryVectors, List<List<float[]>> searchResults, float[][] groundTruthValues, int k, SpaceType spaceType){
+        ArrayList<Float> recalls = new ArrayList<>();
+        float dist = 0.0F;
+
+        for(int i = 0; i < queryVectors.length; i++){
+            float recallVal = 0.0F;
+            for(int j = 0; j < searchResults.get(i).size(); j++){
+                if(spaceType != null && "l2".equals(spaceType.getValue())){
+                    dist = KNNScoringUtil.l2Squared(queryVectors[i], searchResults.get(i).get(j));
```
[the remark this answers is not in the document]
Comment (naveentatikonda) [reply src/test/java/org/opensearch/knn/TestUtils.java:125]: Ack

**Review by martin-gaievski — APPROVED** [review]

File: src/test/java/org/opensearch/knn/bwc/KNNBackwardsCompatibilityIT.java [hunk 30]
```diff
@@ -103,6 +117,11 @@ public void testBackwardsCompatibility() throws Exception {
                     knnWarmup(Collections.singletonList(testIndexName));
                     assertEquals(graphCountBefore + 1, getTotalGraphsInCache());
 
+                    addDocs(testIndex_Recall, testField_Recall, dimensions_Recall, docCount, true);
```
Comment (jmazanec15) [inline src/test/java/org/opensearch/knn/bwc/KNNBackwardsCompatibilityIT.java:120]: Can you summarize what is happening here? This test case is starting to get fairly complex.

**Review by jmazanec15 — COMMENTED** [review]
There seems to be some duplicate code in RecallIT and KNNBackwardsCompatibilityIT. Why is this necessary?

File: src/test/java/org/opensearch/knn/bwc/KNNBackwardsCompatibilityIT.java [hunk 31]
```diff
@@ -194,4 +226,41 @@ private String getUri(ClusterType clusterType) {
...
+       List<Set<String>> groundTruthValues = TestUtils.computeGroundTruthValues(indexVectors, queryVectors, spaceType, k);
+       List<List<String>> searchResults = getSearchResults(testIndex, testField, queryVectors, k);
+       return TestUtils.calculateRecallValue(searchResults, groundTruthValues, k);
+   }
+
+    private void addDocs(String testIndex, String testField, int dimensions, int docCount, boolean isStandard) throws Exception {
+        createKnnIndex(testIndex, getKNNDefaultIndexSettings(), createKnnIndexMapping(testField, dimensions));
+
+        updateClusterSettings(KNN_ALGO_PARAM_INDEX_THREAD_QTY, 2);
+        updateClusterSettings(KNN_MEMORY_CIRCUIT_BREAKER_ENABLED, true);
```
[the remark this answers is not in the document]
Comment (naveentatikonda) [reply src/test/java/org/opensearch/knn/bwc/KNNBackwardsCompatibilityIT.java:242]: We are updating cluster settings, updating index thread qty to 2 to speed up data ingestion. But, sometimes if there are more number of docs, it leads to heap out of memory exception. To fix it, I have enabled circuit breaker.

File: src/test/java/org/opensearch/knn/bwc/KNNBackwardsCompatibilityIT.java [hunk 32]
```diff
@@ -194,4 +226,41 @@ private String getUri(ClusterType clusterType) {
...
+
+        bulkAddKnnDocs(testIndex, testField, TestUtils.getIndexVectors(docCount, dimensions, isStandard), docCount);
+    }
+
+    private List<List<String>> getSearchResults(String testIndex, String testField, float[][] queryVectors, int k) throws IOException {
```
[the remark this answers is not in the document]
Comment (naveentatikonda) [reply src/test/java/org/opensearch/knn/bwc/KNNBackwardsCompatibilityIT.java:247]: We can't create another abstract class because again we can't extend 2 classes(abstract class and KNNRestTestCase). 

For now, as discussed offline will move getSearchResults to knnRestTestCase.

**Review by martin-gaievski — APPROVED** [review]

**Review by jmazanec15 — APPROVED** [review]
